File: backend-flask/services/queue_service.py
# Queue service for code analysis jobs
# For Vercel deployment, jobs run synchronously
# For production with longer timeouts, use Celery or similar

import threading
from services.analysis_service import CodeAnalyzer
from models.models import Project, AnalysisResult, ProjectStatus
from app import db
import uuid


def add_analysis_job(project_id, local_path):
    """Add analysis job to queue and process it"""
    print(f'Analysis job queued for project {project_id} at {local_path}')

    # For Vercel, run in background thread (limited to 10s execution time on hobby plan)
    # For longer analyses, consider external queue service
    thread = threading.Thread(target=process_analysis_job, args=(project_id, local_path))
    thread.daemon = True
    thread.start()


def process_analysis_job(project_id, local_path):
    """Process code analysis job"""
    try:
        print(f'Starting analysis for project {project_id}')

        # Update project status
        project = Project.query.get(project_id)
        if not project:
            print(f'Project {project_id} not found')
            return

        project.status = ProjectStatus.ANALYZING
        db.session.commit()

        # Run analysis
        analyzer = CodeAnalyzer(local_path)
        results = analyzer.analyze_project()

        # Store results in database
        for file_result in results['results']:
            analysis_result = AnalysisResult(
                id=str(uuid.uuid4()),
                project_id=project_id,
                file_path=file_result['file_path'],
                language=file_result['language'],
                issues=file_result['issues'],
                suggestions=file_result['suggestions'],
                metrics=file_result['metrics']
            )
            db.session.add(analysis_result)

        # Update project status
        project.status = ProjectStatus.COMPLETED
        project.file_count = results['files_analyzed']
        db.session.commit()

        print(f'Analysis completed for project {project_id}: {results["files_analyzed"]} files, {results["total_issues"]} issues')

    except Exception as e:
        print(f'Error processing analysis job: {e}')
        try:
            project = Project.query.get(project_id)
            if project:
                project.status = ProjectStatus.FAILED
                db.session.commit()
        except Exception as db_error:
            print(f'Error updating project status: {db_error}')

File: backend/src/services/queue.service.ts
import Queue from 'bull'
import Redis from 'ioredis'
import { prisma } from '../utils/prisma.js'
import { GitService } from './git.service.js'
import { AnalysisService } from './analysis.service.js'
import { AIService } from './ai.service.js'

const redisUrl = process.env.REDIS_URL || 'redis://localhost:6379'

// Create Redis client for Bull
const redisClient = new Redis(redisUrl, {
  maxRetriesPerRequest: null,
})

// Create analysis queue
export const analysisQueue = new Queue('analysis', {
  redis: redisUrl,
  defaultJobOptions: {
    attempts: 3,
    backoff: {
      type: 'exponential',
      delay: 2000,
    },
    removeOnComplete: 100,
    removeOnFail: 100,
  },
})

export interface AnalysisJob {
  projectId: string
  localPath: string
}

export async function setupJobQueue() {
  const gitService = new GitService()
  const analysisService = new AnalysisService()
  const aiService = new AIService()

  // Process analysis jobs
  analysisQueue.process(async (job) => {
    const { projectId, localPath } = job.data as AnalysisJob

    console.log(`Starting analysis for project ${projectId}`)

    try {
      // Update project status
      await prisma.project.update({
        where: { id: projectId },
        data: { status: 'ANALYZING' },
      })

      // Get all code files
      const codeFiles = await gitService.getAllCodeFiles(localPath)
      console.log(`Found ${codeFiles.length} code files`)

      // Analyze each file
      const results = []
      for (const filePath of codeFiles) {
        const relativePath = filePath.replace(localPath + '/', '')
        const language = gitService['getLanguageFromPath'](filePath)

        console.log(`Analyzing ${relativePath}`)

        // Read file content
        const content = await gitService.getFileContent(filePath)

        // Run static analysis
        const { issues, metrics } = await analysisService.analyzeFile(
          filePath,
          language
        )

        // Generate AI suggestions (only for important files)
        let suggestions = []
        if (issues.length > 0 || metrics.complexity > 10) {
          try {
            suggestions = await aiService.generateSuggestions(
              filePath,
              content,
              language
            )
          } catch (error) {
            console.error(`Failed to generate AI suggestions for ${relativePath}:`, error)
          }
        }

        // Store analysis result
        const result = await prisma.analysisResult.create({
          data: {
            projectId,
            filePath: relativePath,
            language,
            issues: JSON.stringify(issues),
            suggestions: JSON.stringify(suggestions),
            metrics: JSON.stringify(metrics),
          },
        })

        results.push(result)

        // Update progress
        job.progress((results.length / codeFiles.length) * 100)
      }

      // Update project status
      await prisma.project.update({
        where: { id: projectId },
        data: {
          status: 'COMPLETED',
          fileCount: codeFiles.length,
        },
      })

      console.log(`Completed analysis for project ${projectId}`)
      return { success: true, filesAnalyzed: codeFiles.length }
    } catch (error: any) {
      console.error(`Analysis failed for project ${projectId}:`, error)

      // Update project status to failed
      await prisma.project.update({
        where: { id: projectId },
        data: { status: 'FAILED' },
      })

      throw error
    }
  })

  // Event handlers
  analysisQueue.on('completed', (job, result) => {
    console.log(`Job ${job.id} completed:`, result)
  })

  analysisQueue.on('failed', (job, error) => {
    console.error(`Job ${job?.id} failed:`, error)
  })

  analysisQueue.on('progress', (job, progress) => {
    console.log(`Job ${job.id} progress: ${progress}%`)
  })

  console.log('✅ Job queue initialized')
}

export async function addAnalysisJob(projectId: string, localPath: string) {
  return await analysisQueue.add({ projectId, localPath })
}
